### DISCUSSION TITLE
Help needed: Considering SLINT for Our In-Vehicle Infotainment (IVI) System

### DISCUSSION THREAD

**Discussion by jfbaro** [slint-ui/slint#7121] [Q&A]
Hi there,

We’re exploring SLINT as the UI framework for our next-gen automotive platform, covering infotainment, clusters, HUBs, and secondary screens (if present in the car model). We’d love your insights on why SLINT might be the best fit compared to Qt, Flutter, Kotlin Multi-Platform, or React Native. Our main focus areas include:

1. Modular Development & Micro-Frontends: How well does SLINT support a modular architecture where multiple teams can develop separate features independently, each able to deploy incremental OTA updates without interfering with each other’s code?
2. Performance & Efficiency: Is SLINT genuinely lighter on resources, and can it handle the heavy lifting of a complex IVI system without straining hardware? How does it compare to QT in this regard?
3. Over-the-Air (OTA) Updates: How does SLINT handle incremental updates, including the ability to roll back easily if something goes wrong? Signed updates, btw.
4. Backend Integration: Does SLINT offer first-class support for Rust and/or JS/TypeScript for in-vehicle logic?
5. Developer Experience & Productivity: Will our teams find SLINT intuitive enough to quickly ship robust features, maintain quality with automated testing, and match or exceed their productivity in other frameworks?
6. Branding & Personalization: Can SLINT handle dynamic themes, skins, and a variety of screen shapes and sizes, including round or non-standard displays?
7. Deployment Strategy: Is it realistic to pre-install a minimal core and then dynamically load features and themes as needed, according to the CAR model that driver is onboarding on?
8. Security & Certifications: How does SLINT address security concerns for connected automotive environments? Are there recommended best practices or certifications?
9. Long-Term Support & Roadmap: What ensures SLINT won’t be abandoned? Are there **pro services**, a responsive community, or a clear product roadmap to keep it evolving?
10. Licensing & Contributions: How flexible is the licensing model for automotive use, and can we contribute features or extensions easily if we need something that isn’t available yet?
11. Scenarios & Strengths: In which scenarios does SLINT truly excel compared to its competitors, and where might we hit its limits for modern automotive use?
12. We’d also love to hear about any real-world automotive cases, and whether teams can be as productive with SLINT as they are with more established frameworks like Flutter.

Thanks for your time and insights! If this works out, I’ll owe you a coffee, or maybe even a car-shaped cookie.

**Comment by Vadoola** [comment]
Lot of questions. I'm not a member of the Slint team, and can't really answer most of them, but I might be able to provide an insight into some of them until a member of the Slint team comes and gives you better answers.

> 2. Performance & Efficiency: Is SLINT genuinely lighter on resources, and can it handle the heavy lifting of a complex IVI system without straining hardware? How does it compare to QT in this regard?

The way I understand it, if using Rust or C++ the, Slint code is compiled down into Rust/C++, so it should be pretty performant, and the Slint dev's have some demo's showing it can be used one some resource constrained devices. I've personally never done embedded development, so I can't really add anything here. In compared to Qt, if using Qt widgets, since it's raw C++ I imagine it would be similar, but I don't have any comparisons. Compared to Qml, which I imagine would be better suited for an IVI, Qml is interpreted so, I'm pretty sure it's slower.  You mention in question 4 about JS/Typescript, I'm pretty sure with JS the Slint code is interpreted, so I assume it would be slower, than when compiled to Rust or C++, but I don't have anything to back this up....this is stretching my knowledge here.

> 3. Over-the-Air (OTA) Updates: How does SLINT handle incremental updates, including the ability to roll back easily if something goes wrong? Signed updates, btw.

Since Slint just focuses on being a UI toolkit and DSL (unlike Qt which is an entire ecosystem), as far as I understand there would be no direct support for OTA in Slint, but also no reason it shouldn't work. I know the project has partnerships with Toradex, and I think I saw something once about Slint project running on top of Toradex's Torizon distro and supporting OTA, but I can't say for certain.

> 6. Branding & Personalization: Can SLINT handle dynamic themes, skins, and a variety of screen shapes and sizes, including round or non-standard displays?

If using some of the builtin components such as the standard Button, you might be limited to the built in styles (Fluent, Material, etc). I think they are working on adding more styling options, but they aren't released yet. You can create your own buttons and such from scratch and make them fully theme-able. My project [Tomotroid](https://github.com/Vadoola/Tomotroid) is a good example of this. All the buttons, and the window frame are custom made using Slint, and are fully theme-able, with the colors being swapped at runtime from JSON theme files.

> 7. Deployment Strategy: Is it realistic to pre-install a minimal core and then dynamically load features and themes as needed, according to the CAR model that driver is onboarding on?

Referring to what I said in response to question 6, loading themes as needed should be easy. In Tomotroid I only change the colors, and do some colorization, and color swapping in some SVG files that are bundled in the executable, but I see no reason these images couldn't be loaded at runtime either.

In regards to loading features dynamically, this one I'm less sure about. Since if using a Rust/C++ backend the Slint code is compiled down into native code, loading features dynamically is probably a bit more tricky, and I haven't tried it. If you know of the feature in advance, and and code it up and then just enable/disable at runtime that's probably easy, but adding a new feature after the fact, I don't know. You can use an interpreter for Slint from within Rust or C++, or there might be a better way. Probably best to wait for an answer from the Slint devs on this one.

Hopefully his info helps you out a little bit while waiting for a better answer

**Comment by aurindam** [comment]
@jfbaro tried to contact you via email. Could you please a book a call at a time slot suitable to you - https://cal.slint.dev/aurindam? Thanks.
